### PR TITLE
Lexer refactor suggestion

### PR DESCRIPTION
**Pull request by owickstrom**
Extracts a `lex_doc_string` from the lexer. Could do the same thing with other tokens.

### CODE REVIEW

File: rust/src/lexer.rs [hunk 1]
```diff
@@ -61,60 +63,66 @@ impl<'a, 'b> Lexer<'a, 'b> {
...
-      ch = self.iterator.next()?;
-      position = self.iterator.position.previous_column();
-    }
-    if ch.is_ascii_digit() {
+  }
+  fn lex_int_lit(&mut self) -> Option<LexItem<'a>> {
+    let negation = if self.next_equals('-')? {
+      // Skip the peeked dash.
+      self.iterator.next()?;
+      Some(self.iterator.position.previous_column())
```
Comment (liamoc) [inline rust/src/lexer.rs]: Note that we must not parse subtractions as negations.. e.g. 3 - 4 must still parse OK. That's why I require in the original code that the - must be followed immediately by a digit for negation to apply.

File: rust/src/lexer.rs [hunk 2]
```diff
@@ -325,7 +357,7 @@ mod tests {
   fn lex_doc() {
     expect_lex(
       vec!["/// hello"],
-      vec![(0, 0, Symbol::Doc(String::from(" hello")))],
+      vec![(0, 3, Symbol::Doc(String::from(" hello")))],
```
Comment (liamoc) [inline rust/src/lexer.rs]: I think the position of the docstring should be 0 here not 3.

**Review by liamoc — CHANGES_REQUESTED** [review]

File: rust/src/lexer.rs (hunk 1 above)
Comment (owickstrom) [reply rust/src/lexer.rs]: All right, I'll add tests for subtraction and negation and make sure it works as expected.

File: rust/src/lexer.rs (hunk 2 above)
Comment (owickstrom) [reply rust/src/lexer.rs]: Ack, fixing.

**Comment by owickstrom** [comment]
> I think this will break when trying to parse text after multiple comments e.g.
> 
> ```
> //foo
> //foo2
> bar
> ```

Ah yes, I moved the `lex_doc` from the loop, that's not great...

**Pull request merged**